core/src/main/java/org/elasticsearch/gateway/MetaStateService.java

```diff
@@ -108,11 +104,36 @@ MetaData loadFullState() throws Exception {
      * Loads the index state for the provided index name, returning null if doesn't exists.
      */
     @Nullable
-    IndexMetaData loadIndexState(String index) throws IOException {
+    IndexMetaData loadIndexState(Index index) throws IOException {
         return indexStateFormat.loadLatestState(logger, nodeEnv.indexPaths(index));
     }
 
     /**
+     * Loads all indices states available on disk
+     */
+    List<IndexMetaData> loadIndicesStates(Predicate<String> excludeIndexPathIdsPredicate) throws IOException {
+        List<IndexMetaData> indexMetaDataList = new ArrayList<>();
+        for (String indexFolderName : nodeEnv.availableIndexFolders()) {
+            if (excludeIndexPathIdsPredicate.test(indexFolderName)) {
+                continue;
+            }
+            IndexMetaData indexMetaData = indexStateFormat.loadLatestState(logger,
+                nodeEnv.resolveIndexFolder(indexFolderName));
+            if (indexMetaData != null) {
+                final String indexPathId = indexMetaData.getIndex().getPathIdentifier();
+                if (indexFolderName.equals(indexPathId)) {
+                    indexMetaDataList.add(indexMetaData);
+                } else {
+                    logger.info("[{}] invalid folder name, rename to [{}], ignoring", indexFolderName, indexPathId);
```

Comment:
this feels too lenient to me. At the very least we should add a parameter indicating whether it's OK to be lenient (dangling indices detection ) and be strict (throw an exception on node start up).